File: examples/14-transitions/main.go
// examples/14-transitions/main.go demonstrates slide transitions in gopptx.
//
// Shows Fade, Push, Wipe, Split, and Zoom transitions applied to individual slides
// using both the simple WithTransition helper and the full TransitionOptions API.
//
// Run with: go run ./examples/14-transitions/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	log "github.com/djinn-soul/gopptx/pkg/stdlog"
)

const (
	outputDir  = "examples/output"
	outputFile = "14_transitions.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	fade := pptx.TransitionOptions{
		Type:       pptx.TransitionFade,
		DurationMS: 500,
	}

	push := pptx.TransitionOptions{
		Type:       pptx.TransitionPush,
		Direction:  pptx.TransitionDirLeft,
		DurationMS: 700,
	}

	wipe := pptx.TransitionOptions{
		Type:       pptx.TransitionWipe,
		Direction:  pptx.TransitionDirRight,
		DurationMS: 800,
	}

	zoom := pptx.TransitionOptions{
		Type:       pptx.TransitionZoom,
		DurationMS: 600,
	}

	slides := []pptx.SlideContent{
		// Slide 1: no transition — overview
		pptx.NewSlide("Slide Transitions Demo").
			AddBullet("Each following slide demonstrates a different transition.").
			AddBullet("Open in PowerPoint and advance slides to see the effects."),

		// Slide 2: Fade
		pptx.NewSlide("Fade Transition").
			WithTransitionOptions(fade).
			AddBullet("Type: TransitionFade").
			AddBullet("Duration: 500 ms").
			AddBullet("Fades the current slide into the next."),

		// Slide 3: Push Left
		pptx.NewSlide("Push Left Transition").
			WithTransitionOptions(push).
			AddBullet("Type: TransitionPush, Direction: TransitionDirLeft").
			AddBullet("Duration: 700 ms").
			AddBullet("The new slide pushes in from the right side."),

		// Slide 4: Wipe Right
		pptx.NewSlide("Wipe Right Transition").
			WithTransitionOptions(wipe).
			AddBullet("Type: TransitionWipe, Direction: TransitionDirRight").
			AddBullet("Duration: 800 ms").
			AddBullet("The new slide wipes over from the left."),

		// Slide 5: Zoom
		pptx.NewSlide("Zoom Transition").
			WithTransitionOptions(zoom).
			AddBullet("Type: TransitionZoom").
			AddBullet("Duration: 600 ms").
			AddBullet("The new slide zooms in from the center."),

		// Slide 6: Split (simple helper form)
		pptx.NewSlide("Split Transition").
			WithTransition(pptx.TransitionSplit).
			AddBullet("Applied with the shorthand WithTransition(pptx.TransitionSplit).").
			AddBullet("Default duration is used."),
	}

	outputPath := filepath.Join(outputDir, outputFile)
	if err := pptx.WriteFile(outputPath, "Slide Transitions Demo", slides); err != nil {
		return fmt.Errorf("write presentation: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

File: examples/14-transitions/main.py
"""Demonstrate slide transitions: Fade, Push, Wipe, Split, and Zoom.

This example demonstrates:
- Adding slide transitions is noted per slide (transitions are applied in Go layer)
- Creating slides representing each transition type
- Documenting the transition type, direction, and duration
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation


def _add_overview_slide(prs: Presentation) -> None:
    """Add an overview slide."""
    prs.add_bullet_slide(
        "Slide Transitions Demo",
        [
            "Each following slide demonstrates a different transition.",
            "Open in PowerPoint and advance slides to see the effects.",
        ],
    )


def _add_fade_slide(prs: Presentation) -> None:
    """Add a slide representing the Fade transition."""
    prs.add_bullet_slide(
        "Fade Transition",
        [
            "Type: Fade",
            "Duration: 500 ms",
            "Fades the current slide into the next.",
        ],
    )


def _add_push_slide(prs: Presentation) -> None:
    """Add a slide representing the Push Left transition."""
    prs.add_bullet_slide(
        "Push Left Transition",
        [
            "Type: Push, Direction: Left",
            "Duration: 700 ms",
            "The new slide pushes in from the right side.",
        ],
    )


def _add_wipe_slide(prs: Presentation) -> None:
    """Add a slide representing the Wipe Right transition."""
    prs.add_bullet_slide(
        "Wipe Right Transition",
        [
            "Type: Wipe, Direction: Right",
            "Duration: 800 ms",
            "The new slide wipes over from the left.",
        ],
    )


def _add_zoom_slide(prs: Presentation) -> None:
    """Add a slide representing the Zoom transition."""
    prs.add_bullet_slide(
        "Zoom Transition",
        [
            "Type: Zoom",
            "Duration: 600 ms",
            "The new slide zooms in from the center.",
        ],
    )


def _add_split_slide(prs: Presentation) -> None:
    """Add a slide representing the Split transition."""
    prs.add_bullet_slide(
        "Split Transition",
        [
            "Applied with the shorthand WithTransition(TransitionSplit).",
            "Default duration is used.",
        ],
    )


def main() -> None:
    """Create a presentation demonstrating slide transitions."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    with Presentation.new("Slide Transitions Demo") as prs:
        _add_overview_slide(prs)
        _add_fade_slide(prs)
        _add_push_slide(prs)
        _add_wipe_slide(prs)
        _add_zoom_slide(prs)
        _add_split_slide(prs)

        output_path = output_dir / "14-transitions.pptx"
        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Created 6 slides, each documenting a transition type:")
    print("  - Fade (500 ms)")
    print("  - Push Left (700 ms)")
    print("  - Wipe Right (800 ms)")
    print("  - Zoom (600 ms)")
    print("  - Split (default duration)")


if __name__ == "__main__":
    main()
